Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp

```diff
@@ -0,0 +1,277 @@
+          this, &IOWindow::OnRangeChanged);
+  connect(m_range_slider, static_cast<void (QSlider::*)(int value)>(&QSlider::valueChanged), this,
+          &IOWindow::OnRangeChanged);
+}
+
+void IOWindow::AppendSelectedOption(const QString& prefix)
+{
+  if (m_option_list->currentItem() == nullptr)
+    return;
+
+  m_expression_text->insertPlainText(
+      MappingCommon::GetExpressionForControl(m_option_list->currentItem()->text(), &m_devq,
+                                             &m_controller->default_device)
+          .prepend(prefix));
+}
+
+void IOWindow::OnDeviceChanged(const QString& device)
+{
+  m_devq.FromString(device.toStdString());
+  UpdateOptionList();
+}
+
+void IOWindow::OnDialogButtonPressed(QAbstractButton* button)
+{
+  if (button == m_clear_button)
+  {
+    m_expression_text->clear();
+    return;
+  }
+  else
+  {
+    m_reference->expression = m_expression_text->toPlainText().toStdString();
+  }
+
+  if (button != m_apply_button)
+    accept();
+}
+
+void IOWindow::OnDetectButtonPressed()
+{
+  if (m_block)
+    return;
+
+  m_block = true;
+  std::thread([this] {
+    auto* btn = m_type == IOWindow::Type::Input ? m_detect_button : m_test_button;
+    const auto old_label = btn->text();
+
+    btn->setText(QStringLiteral("..."));
+
+    ciface::Core::Device::Control* const ctrl =
+        m_reference->Detect(5000, g_controller_interface.FindDevice(m_devq).get());
+
+    btn->setText(old_label);
+
+    if (ctrl)
+    {
+      const auto name = QString::fromStdString(ctrl->GetName());
+      const auto list = m_option_list->findItems(name, Qt::MatchFixedString);
+
+      if (list.size() > 0)
+        m_option_list->setCurrentItem(list[0]);
+    }
+    m_block = false;
+  }).detach();
+}
+
+void IOWindow::OnRangeChanged(int value)
+{
+  m_reference->range = static_cast<double>(value) / SLIDER_TICK_COUNT;
+  m_range_spinbox->setValue(m_reference->range * SLIDER_TICK_COUNT);
+  m_range_slider->setValue(m_reference->range * SLIDER_TICK_COUNT);
+}
+
+void IOWindow::UpdateOptionList()
+{
+  if (m_block)
+    return;
+
+  m_option_list->clear();
+
+  const auto device = g_controller_interface.FindDevice(m_devq);
+  if (m_reference->IsInput())
+  {
+    for (const auto* input : device->Inputs())
+    {
+      m_option_list->addItem(QString::fromStdString(input->GetName()));
+    }
+  }
+  else
+  {
+    for (const auto* output : device->Outputs())
+    {
+      m_option_list->addItem(QString::fromStdString(output->GetName()));
+    }
+  }
+}
+
+void IOWindow::UpdateDeviceList()
+{
+  m_block = true;
+  m_devices_combo->clear();
+
+  const bool paused = Core::PauseAndLock(true);
+
+  g_controller_interface.RefreshDevices();
+  m_controller->UpdateReferences(g_controller_interface);
+  m_controller->UpdateDefaultDevice();
+
+  const auto default_device = m_controller->default_device.ToString();
+
+  m_devices_combo->addItem(QString::fromStdString(default_device));
```

Comment:
Can you put a comment on all of this to explain why the default device is added manually?